### PR TITLE
Small refactor, adding a list of all kinds to BuildContext

### PR DESCRIPTION
**Pull request by Ekleog**
Involved in https://github.com/rust-lang/cargo/pull/9030

cc @alexcrichton

### CODE REVIEW

**Comment by Ekleog** [comment]
Hmm the test failure here looks very dubious to me:
```
#[cargo_test]
#[cfg(all(target_arch = "x86_64", target_os = "linux", target_env = "gnu"))]
fn cfg_looks_at_rustflags_for_target() {
```
yet it happens on “CI / test (ubuntu-latest, beta, i686-unknown-linux-gnu) (pull_request)” ; which is thus not `x86_64`? (Locally, the test passes)

For `install_target_native`, things are less clear, but given the above and the fact that it passes locally I guess I'll wait for someone to help me about the above at least

Finally, for the test that starts with `aaa_`, I've never managed to make it pass locally, even with cargo master, plus it's written that it's ok for this test to fail, so I guess things are okay-ish.

TL;DR: locally (on x86_64) all tests pass, this looks like it may be due to it being an i686 runner, but OTOH one of the tests is supposed to be disabled for non-x86_64, so I'll say this is probably CI misbehaving. Does this assumption make sense, or do I actually have things to fix?

**Comment by Ekleog** [comment]
(For people who got a notification with my previous message: it was stupid, and it looks like one `git push` I thought I had made before opening the PR didn't actually happen. CI will probably pass now that I've re-pushed the right branch.)

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!
